docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,485 @@
+
+### Select
+
+The Select operator (`| select`) allows you to select and reduce field that should be returned.
+
+If you don't provide a select operator, all fields parsed will be returned by default. This means all fields from a log line if you're using implicit extraction. However if you used explicit extraction, only those fields will be returned.
+
+For example this query:
+
+```logql
+{job="prod/query-frontend"} | logfmt | sort duration desc | limit 25 | select query duration
+```
+
+is the same as this one
+
+```logql
+{job="prod/query-frontend"} | logfmt query duration | sort duration desc | limit 25
+```
+
+because the field extraction was explicit, only those fields (`query` and `duration`) are outputted.
+
+This is why select is important to use for large queries. Select will allows users to remove data (column) in their logs that are useless, repetitive and not interesting for the current investigation.
+
+The select operation will also support the [go templating language](https://golang.org/pkg/text/template/) when supplied with string argument and not fields.
+
+Fields extracted will be available under the `.FieldName` and `$FieldName` notation in the template.
+
+This way users can format each field to be more digestible. Returning a payload size  field with value of `12321421230123` is less meaningful than `12TB`.
+
+Example:
+
+```logql
+{job="prod/query-frontend"} | logfmt | select payload_size="{{ .size | humanize }}" level="{{ $level | ToUpper}}"
+```
+
+Renaming is allowed with or without templates, and is optional. An autogenerated field name will be used if it can't be inferred. (`field1`,`field2`, etc...)
+
+__In this current stance, select is not greedy meaning if you want to rename or template a single field, you'll have to declare all other fields to keep them.__
+
+In the future we might consider a `exclude` operator to take all fields but those passed as parameter. (`{job="prod/query-frontend"} | logfmt | exclude msg`)
+
+### Limit
+
+The limit operator (`| limit x`) is a new operator to overwrite the query string limit, this can be handy for building dashboard with different limits but also when writing and testing queries, you could start to experiment with a low limit to increase the feedback loop.
+
+If the limit operator is missing in the expression, the limit from the query string is used.
+
+The limit operator can be applied to a log field  as well as a log stream.
+
+For example all of the above are correct.
+
+```logql
+{job="prod/query-frontend"} | logfmt query duration | sort duration desc | limit 25
+{job="prod/query-frontend"} | limit 10
+{job="prod/query-frontend"} |= "foo" | limit 20
+```
+
+Using limit operator with metric queries will result into an error. (e.g `rate({job="prod/query-frontend"} | limit 10[5m])`)
+
+### Uniq
```

Comment:
What's the use case for this? You mentioned "This is really useful to reduce noise applications might generate by failing constantly" but I would consider more useful the filter expression `!= "noise log"` unless the `uniq` remove duplicates only when they're consecutive (I think this is unclear from the proposal).

Reply:
It doesn't remove only consecutive simply all duplicates lines.

The use case is exactly what you describe with the `!= "noise log"`, the problem I've seen with this solution with my own experience is that you endup writing this:

`{app="noisy"} != "not that one" != "nope nope" != "still not" != "and no"`

Reply:
Other uses cases I've had for `uniq` in the past are something like this `cat log.txt | cut -d " " -f 2 | sort | uniq` where I have a lot like:
```
2020-03-20Z00:00:00 slim-bean bunch of other stuff...
```
Something similar to what you might see in an apache access log where I wanted to get a list of unique users in the logs.

Similarly i've done this with IP addresses, etc